doc/src/extras.rst

```diff
@@ -141,6 +141,262 @@ Logging cursor
 
 .. autoclass:: MinTimeLoggingCursor
 
+Replication cursor
+^^^^^^^^^^^^^^^^^^
+
+.. autoclass:: ReplicationConnection
+
+   This connection factory class can be used to open a special type of
+   connection that is used for streaming replication.
+
+   Example::
+
+       from psycopg2.extras import ReplicationConnection, REPLICATION_PHYSICAL, REPLICATION_LOGICAL
+       conn = psycopg2.connect(dsn, connection_factory=ReplicationConnection)
+       cur = conn.cursor()
+
+.. seealso::
+
+   - PostgreSQL `Replication protocol`__
+
+   .. __: http://www.postgresql.org/docs/current/static/protocol-replication.html
+
+.. autoclass:: ReplicationCursor
+
+    .. method:: identify_system()
+
+       Execute ``IDENTIFY_SYSTEM`` command of the streaming replication
+       protocol and return the result as a dictionary.
+
+       Example::
+
+           >>> cur.identify_system()
+           {'timeline': 1, 'systemid': '1234567890123456789', 'dbname': 'test', 'xlogpos': '0/1ABCDEF'}
+
+    .. method:: create_replication_slot(slot_type, slot_name, output_plugin=None)
+
+       Create streaming replication slot.
+
+       :param slot_type: type of replication: either `REPLICATION_PHYSICAL` or
+                         `REPLICATION_LOGICAL`
+       :param slot_name: name of the replication slot to be created
+       :param output_plugin: name of the logical decoding output plugin to use
+                             (logical replication only)
+
+       Example::
+
+           cur.create_replication_slot(REPLICATION_LOGICAL, "testslot", "test_decoding")
+
+    .. method:: drop_replication_slot(slot_name)
+
+       Drop streaming replication slot.
+
+       :param slot_name: name of the replication slot to drop
+
+       Example::
+
+           cur.drop_replication_slot("testslot")
+
+    .. method:: start_replication(slot_type, slot_name=None, writer=None, start_lsn=None, timeline=0, keepalive_interval=10, options=None)
+
+       Start a replication stream.  On non-asynchronous connection, also
+       consume the stream messages.
+
+       :param slot_type: type of replication: either `REPLICATION_PHYSICAL` or
+                         `REPLICATION_LOGICAL`
+       :param slot_name: name of the replication slot to use (required for
+                         logical replication)
+       :param writer: a file-like object to write replication messages to
```

Comment:
This is IMO the wrong API here. A file-like object is not suitable for COPY BOTH, because you may need clearly delimited messages. Sometimes it's not possible to interpret a stream of binary results *without* knowing the message boundaries and lengths from `CopyData` messages.

IMO you should either have a listener pattern here, where you call an object that exposes a method like `copy_message(msgbytes)`, or you should return a generator that does something like yields byte strings (for each message) until generation end. The generator is strongly preferable IMO because it lets the client control the protocol flow and inject feedback messages as it goes.

I think that generator should be part of the base psycopg2 cursor and usable from `copy_expert` (or a variant) too, so that you can:

    for csvrow in curs.copy_expert('COPY myhugetable TO stdout', None):
        ... do something with the str object 'csvrow' ...

it's currently super-inefficient to have psycopg2 append the data to a file-like object, only to then parse it out again.

If we need a message-oriented API for COPY BOTH, which we do, why not support it for regular `copy` too?

I do not think this is suitable for merging with a file-like API and should be changed.